### PR TITLE
fix #1151, reduce the memory footprint of each Task

### LINKED ISSUE #1151: scale up to 10, 000 tasks on single four-core machine
In theory, with effecitve flow control, we should be able to scale to as many tasks as we want.

But in reality, it depends on:
1. memory footprint of single task is small enough, so that there is no out of memory.
2. GC pause interval is still under control to avoid remote watching timeout.

### PR DESCRIPTION
**Pull request by clockfly**
[Review on Reviewable](https://reviewable.io/reviews/intel-hadoop/gearpump/1179)

### CODE REVIEW

**Comment by clockfly** [comment]
with this PR, now we reduce the memory overhead per sliding window from 100 bytes to 35 bytes.

**Comment by clockfly** [comment]
There are bunch of other fixes to reduce memory used to metrics, kryo.

**Comment by manuzhang** [comment]
subscription spec failed 

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/intel-hadoop/gearpump/1179)

File: streaming/src/main/scala/org/apache/gearpump/streaming/executor/TaskLauncher.scala [hunk 1]
```diff
@@ -19,16 +19,18 @@
...
+import org.apache.gearpump.serializer.KryoPool
 import org.apache.gearpump.streaming.ProcessorDescription
 import org.apache.gearpump.streaming.executor.TaskLauncher.TaskArgument
 import org.apache.gearpump.streaming.task.{Subscriber, TaskWrapper, TaskActor, TaskUtil, TaskContextData, TaskId}
 import org.apache.gearpump.streaming.util.ActorPathUtil
 import org.apache.gearpump.util.Constants
-import akka.actor.Actor
 
 trait ITaskLauncher {
+  def setKryoPool(pool: KryoPool): Unit
```
Comment (huafengw) [inline streaming/src/main/scala/org/apache/gearpump/streaming/executor/TaskLauncher.scala]: This interface maybe not really make sense in ITaskLauncher trait.

**Comment by manuzhang** [comment]
Review status: 0 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala, line 44 \[r1\]](https://reviewable.io:443/reviews/intel-hadoop/gearpump/1179#-JuEIf_i1mC18NyRSJNu)** ([raw file](https://github.com/intel-hadoop/gearpump/blob/c144d53f91b00147e5af3e31e5ecbeb319b87b44/streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala#L44)):</sup>
pls. fix indentation

---

<sup>**[streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskControlMessage.scala, line 27 \[r1\]](https://reviewable.io:443/reviews/intel-hadoop/gearpump/1179#-JuEHnWDl_aY4WPA-oie)** ([raw file](https://github.com/intel-hadoop/gearpump/blob/c144d53f91b00147e5af3e31e5ecbeb319b87b44/streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskControlMessage.scala#L27)):</sup>
why not use `Short` for sessionId as well ?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/intel-hadoop/gearpump/1179)

**Comment by manuzhang** [comment]
Review status: 0 of 21 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskManager.scala, line 132 \[r1\]](https://reviewable.io:443/reviews/intel-hadoop/gearpump/1179#-JuELmsvqYODl9mGuFfR)** ([raw file](https://github.com/intel-hadoop/gearpump/blob/c144d53f91b00147e5af3e31e5ecbeb319b87b44/streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/TaskManager.scala#L132)):</sup>
maybe this should no long be named as`StartClock`. maybe simply `Start`

---

<sup>**[streaming/src/main/scala/org/apache/gearpump/streaming/util/KryoSerializerPool.scala, line 9 \[r1\]](https://reviewable.io:443/reviews/intel-hadoop/gearpump/1179#-JuEMyB96FmffsAi8pNp)** ([raw file](https://github.com/intel-hadoop/gearpump/blob/c144d53f91b00147e5af3e31e5ecbeb319b87b44/streaming/src/main/scala/org/apache/gearpump/streaming/util/KryoSerializerPool.scala#L9)):</sup>
where is this used ?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/intel-hadoop/gearpump/1179)

**Comment by manuzhang** [comment]
<sup>**[core/src/main/scala/org/apache/gearpump/transport/netty/Client.scala, line 46 \[r1\]](https://reviewable.io:443/reviews/intel-hadoop/gearpump/1179#-JuEg1TJUNotsrnc8Rkx)** ([raw file](https://github.com/intel-hadoop/gearpump/blob/c144d53f91b00147e5af3e31e5ecbeb319b87b44/core/src/main/scala/org/apache/gearpump/transport/netty/Client.scala#L46)):</sup>
why is `ArrayList` a better choice here ?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/intel-hadoop/gearpump/1179)

**Comment by clockfly** [comment]
@manuzhang 

ArrayBuffer performs probably same as ArrayList.
however, ArrayList support other operations like trimSize, while ArrayBuffer don't.

**Comment by clockfly** [comment]
@manuzhang 

sessionid must be unique. IF we use short for session Id, there will be at most 65536 tasks in the topology.

**Comment by clockfly** [comment]
The remain random UT failure have been addressed by https://github.com/intel-hadoop/gearpump/pull/1164

**Comment by manuzhang** [comment]
+1, please rebase and squash your commits

**Pull request merged**